### PR TITLE
Flatpak

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change
Allows running sandboxed applications.

Closes: #32807

###### How to test

Enable the service by adding `services.flatpak.enable = true;` to your `configuration.nix`

```
flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
flatpak update
flatpak search bustle
flatpak install flathub org.freedesktop.Bustle
flatpak run org.freedesktop.Bustle
```

###### Things done

Installing and running applications works, ~we still need to check whether all the portals and interactions work.~ Portals work as well.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by jtojnar** [comment]
When running `org.freedesktop.Bustle`, I see:

```
flatpak run: Can't get document portal: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.portal.Documents was not provided by any .service files
/lib/libc.so.6: version `GLIBC_2.25' not found (required by /nix/store/5pbbfiwq7lvg6jq8iplfj9gw3jp35g9a-gnutls-3.6.1/lib/libgnutls.so.30)
Failed to load module: /nix/store/cahyfwfc2hznx61sqz2j7p0a6f07n3aw-glib-networking-2.54.1/lib/gio/modules/libgiognutls.so
```

**Comment by jtojnar** [comment]
Weird, if I run `/nix/store/f8algjxyc5dal7jls16rnax7c5ibbhqx-flatpak-0.10.2.1/libexec/xdg-document-portal` manually, it works. For some reason the service does not start automatically.

There is also another message from `org.gnome.Photos`:

```
Tracker-Message: Could not get mtime for 'file:///home/jtojnar/.var/app/org.gnome.Photos/cache/tracker/meta.db': Error when getting information for file “/home/jtojnar/.var/app/org.gnome.Photos/cache/tracker/meta.db”: No such file or directory

(gnome-photos:2): Tracker-WARNING **: Falling back to bus backend, the direct backend failed to initialize: Could not open sqlite3 database:'/home/jtojnar/.var/app/org.gnome.Photos/cache/tracker/meta.db': unable to open database file
```

**Comment by Mounium** [comment]
I used to be able to run flatpak flawlessly with your commits (not sure whether there was an update to them since, I usually just use current trunk with some cherry-picked commits, like yours) but for quite some time now I only get `error: Failed to execute child process ?bwrap? (No such file or directory)` when I try to do `flatpak run` with any package (install, remote works just fine). I've even tried deleting the whole flatpak directory in /var but it didn't help either. Can I help somehow, any useful logs I'm supposed to post?

**Comment by jtojnar** [comment]
@Mounium Sorry about that, I switched from the vendored bubblewrap to the one in Nix and did not check if it works. (I though bubblewrap is used as a library which would be linked correctly but it uses the `bwrap` binary). Should be fixed now.

**Comment by Mounium** [comment]
I have been using it in the last couple of days, most flatpak packages work just fine, awesome work :+1: 
 
One strange package, pulseeffects doesn't seem to work, or at least not in a stable manner (https://github.com/wwmm/pulseeffects/issues/156), as I can get it to work from within the gnome software center, but not from command line, probably due to some environment variable set by the wrapper script on gnome-software.

**Comment by jtojnar** [comment]
I tested the portals with https://github.com/matthiasclasen/portal-test (follow the changes in pull requests) and now they work. The last remaining thing is adding an automated test.

I created a prototype but the `portal-test` does not build because it does not have access to GNOME platform and SDK. We cannot fetch it due to sandboxing, we will need `fetchostree` that will produce a fixed-output derivation like `fetchgit` does.

<details>
<summary>Test prototype</summary>

```patch
--- a/nixos/release.nix
+++ b/nixos/release.nix
@@ -255,6 +255,7 @@
   tests.env = callTest tests/env.nix {};
   tests.ferm = callTest tests/ferm.nix {};
   tests.firefox = callTest tests/firefox.nix {};
+  tests.flatpak = callTest tests/flatpak.nix {};
   tests.firewall = callTest tests/firewall.nix {};
   tests.fleet = callTestOnTheseSystems ["x86_64-linux"] tests/fleet.nix {};
   #tests.gitlab = callTest tests/gitlab.nix {};
--- /dev/null
+++ b/nixos/tests/flatpak-portal-test.patch
@@ -0,0 +1,42 @@
+--- a/flatpak/build.sh
++++ b/flatpak/build.sh
+@@ -1,3 +1,3 @@
+ #!/bin/sh
+ 
+-flatpak-builder --force-clean --ccache --require-changes --repo=repo app org.gnome.PortalTest.json
++flatpak-builder --force-clean --require-changes --repo=repo app org.gnome.PortalTest.json
+--- a/flatpak/org.gnome.PortalTest.json
++++ b/flatpak/org.gnome.PortalTest.json
+@@ -10,23 +10,6 @@
+         "--socket=pulseaudio"
+     ],
+     "modules": [
+-        {
+-            "name": "meson",
+-            "no-autogen": true,
+-            "cleanup": [ "/bin", "/lib" ],
+-            "sources": [
+-                {
+-                    "type": "archive",
+-                    "url": "https://github.com/mesonbuild/meson/releases/download/0.36.0/meson-0.36.0.tar.gz",
+-                    "sha256": "dc087ec40dacb5e256e6ee6467f2d004faf4ef284d3c1ce5e89faa1e16540950"
+-                },
+-                {
+-                    "type": "file",
+-                    "path": "meson-makefile",
+-                     "dest-filename": "Makefile"
+-                }
+-            ]
+-        },
+         {
+             "name": "portal-test",
+             "buildsystem": "meson",
+@@ -34,7 +17,7 @@
+             "sources": [
+                 {
+                     "type": "git",
+-                    "url": "https://github.com/matthiasclasen/permission-viewer.git"
++                    "url": ".."
+                 }
+             ]
+         }
--- /dev/null
+++ b/nixos/tests/flatpak.nix
@@ -0,0 +1,48 @@
+import ./make-test.nix ({ pkgs, ... }:
+
+let
+  portal-test = with pkgs; stdenv.mkDerivation {
+    name = "portal-test";
+    src = fetchFromGitHub {
+      owner = "matthiasclasen";
+      repo = "portal-test";
+      rev = "a2468457cee72b86be51d8a0ee5911c37d6894c2";
+      sha256 = "00c7wcjawh80f9vfqm187hmv325ghl9c6r620iq4bf7q4sqj1ab6";
+    };
+    phases = [ "unpackPhase" "patchPhase" "buildPhase" "installPhase" ];
+    patches = [ ./flatpak-portal-test.patch ];
+    buildInputs = [ flatpak-builder gitMinimal libdwarf ];
+    buildPhase = ''
+      cd flatpak
+      ./build.sh
+    '';
+    installPhase = ''
+      cp -r repo $out/repo
+    '';
+  };
+in {
+  name = "flatpak";
+  meta = {
+    maintainers = pkgs.flatpak.meta.maintainers;
+  };
+
+  machine =
+    { config, pkgs, ... }:
+
+    {
+      imports = [ ./common/x11.nix ];
+      services.flatpak.enable = true;
+    };
+
+  testScript =
+    ''
+      $machine->waitForX;
+      $machine->execute("flatpak remote-add --user --no-gpg-verify portal-test ${portal-test}/repo");
+      $machine->execute("flatpak install --user portal-test org.gnome.PortalTest");
+      $machine->execute("flatpak run org.gnome.PortalTest");
+      $machine->waitForWindow(qr/Portal Test/);
+      $machine->succeed("xwininfo -root -tree | grep 'Portal Test'");
+      $machine->screenshot("screen");
+    '';
+
+})
```
</details>

Sources:
* http://docs.flatpak.org/en/latest/distributing-applications.html
* https://blogs.gnome.org/alexl/2017/02/10/maintaining-a-flatpak-repository/
* https://ostree.readthedocs.io/en/latest/manual/repository-management/
* https://handbrake.fr/docs/en/1.0.0/developer/flatpak-repo.html

**Comment by jtojnar** [comment]
Since the last time, I tried switching to [installed tests](https://github.com/NixOS/nixpkgs/issues/34987) which look more promising than the [test prototype](https://github.com/NixOS/nixpkgs/pull/33371#issuecomment-365149130). Unfortunately, they [try to create a runtime](https://github.com/flatpak/flatpak/blob/0.11.3/tests/make-test-runtime.sh) by copying files from `/usr/lib` etc. This will need to be replaced.

**Comment by peterhoeg** [comment]
There is also ```xdg-desktop-portal-kde``` as part of KDE applications.

**Comment by jtojnar** [comment]
More path patching and adding some dependencies fixed most of the tests:

```
SUMMARY: total=16; passed=9; skipped=0; failed=7; user=279.1s; system=38.6s; maxrss=507736
FAIL: Flatpak/test-bundle.sh.test (Child process exited with code 1)
FAIL: Flatpak/test-extensions.sh.test (Child process exited with code 1)
FAIL: Flatpak/test-run-deltas.sh.test (Child process exited with code 1)
FAIL: Flatpak/test-run-system.sh.test (Child process exited with code 1)
FAIL: Flatpak/test-run-system-deltas.sh.test (Child process exited with code 1)
FAIL: Flatpak/test-bundle-system.sh.test (Child process exited with code 1)
FAIL: Flatpak/test-run.sh.test (Child process exited with code 1)
error: command `gnome-desktop-testing-runner' did not succeed (exit code 2)
```

The seven remaining tests depend on generating a _platform_, which I did by copying a Nix store closure to the export `$DIR` and adding symlinks from FHS in the `$DIR`. This however fails, as `flatpak build-export` seems to discard files outside `$DIR/usr`. I have opened an [upstream issue](https://github.com/flatpak/flatpak/issues/1460) asking about that.

**Comment by jtojnar** [comment]
Thanks to Alex Larsson’s help, all flatpak tests are passing. The only remaining thing is the xdg-desktop-portal test:

```
env XDG_DATA_DIRS=(nix-build -A xdg-desktop-portal.installedTests)/share nix-shell -p gnome-desktop-testing --run 'gnome-desktop-testing-runner'
Running test: xdg-desktop-portal/testdb.test
# random seed: R02S96c908b19a997bd490ee5a1c7b37f3bd
1..3
# Start of db tests
ok 1 /db/open
ok 2 /db/serialize
ok 3 /db/modify
# End of db tests
PASS: xdg-desktop-portal/testdb.test
Running test: xdg-desktop-portal/test-doc-portal.test
# random seed: R02Sd6bf1a368da82934ef9075bc1dc39bf9
outdir: /tmp/xdp-test-2E80FZ
**
xdg-desktop-portal:ERROR:tests/test-doc-portal.c:453:global_setup: assertion failed (error == NULL): GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.portal.Documents was not provided by any .service files (g-dbus-error-quark, 2)
# xdg-desktop-portal:ERROR:tests/test-doc-portal.c:453:global_setup: assertion failed (error == NULL): GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.portal.Documents was not provided by any .service files (g-dbus-error-quark, 2)
cleaning up pid 12915
FAIL: xdg-desktop-portal/test-doc-portal.test (Child process killed by signal 6)
SUMMARY: total=2; passed=1; skipped=0; failed=1; user=0.0s; system=0.0s; maxrss=5008
FAIL: xdg-desktop-portal/test-doc-portal.test (Child process killed by signal 6)
```

For some reason the service is not recognized even though I listed it in `services.dbus.packages`. Anyone has any idea how to fix that?

**Comment by jtojnar** [comment]
For xdg-desktop-portal, it turns out a service file were not being installed, https://github.com/flatpak/xdg-desktop-portal/pull/159 fixed it.

The `tests.flatpak.x86_64-linux` seem to be failing due to a timeout, see the attached log
[tests.flatpak.log.txt](https://github.com/NixOS/nixpkgs/files/1825143/tests.flatpak.log.txt)

**Comment by Mounium** [comment]
Is there a specific reason why isn't it merged? I've used it several times without major problems.

**Comment by jtojnar** [comment]
@Mounium See the comments above. The tests are broken.

**Comment by jtojnar** [comment]
Still the same, need to fix the flatpak test. Help would be appreciated.

**Comment by xeji** [comment]
@jtojnar added a commit that fixes `nixos/tests/flatpak`. Test succeeds now in a local VM but takes about 1 hour, so we'll  have to see how this works out on Hydra.

**Comment by xeji** [comment]
@GrahamcOfBorg build flatpak flatpak-builder xdg-desktop-portal xdg-desktop-portal-gtk

**Comment by xeji** [comment]
The test may time out on ofborg, but let's try it once anyway.

@GrahamcOfBorg test flatpak

**Comment by xeji** [comment]
Timed out as expected.

**Comment by jtojnar** [comment]
Thanks to @xeji, the tests now work. I updated the package and it is ready for review.

File: nixos/tests/flatpak.nix [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  meta = {
+    maintainers = pkgs.flatpak.meta.maintainers;
+  };
+
+  machine = { config, pkgs, ... }: {
+    imports = [ ./common/x11.nix ];
+    services.xserver.desktopManager.gnome3.enable = true; # TODO: figure out minimal environment where the tests work
+    services.flatpak.enable = true;
+    environment.systemPackages = with pkgs; [ gnupg gnome-desktop-testing ostree python2 ];
+    virtualisation.memorySize = 2048;
```
Comment (xeji) [inline nixos/tests/flatpak.nix]: If you want the test to run on i686, this should be 2047 due to a memory limitation in `qemu-system-i386`

**Comment by taohansen** [comment]
It'd be nice if users could use Flatpak too. As it is now, any operations having to do with Flatpak have to be done as root making Flatpak support a non-start for me.

`flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo`
`error: Flatpak system operation ConfigureRemote not allowed for user`

**Comment by hedning** [comment]
Have you tried to supply the `--user` flag? I'm at least able to install using my user account, though I can't quite remember if I had to do something special.

**Comment by alexeymuranov** [comment]
On NixOS i am using `flatpak` without `sudo` and without `--user`.

**Pull request merged**